src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -1211,6 +1211,80 @@ ss::future<cluster::init_tm_tx_reply> tx_gateway_frontend::limit_init_tm_tx(
   std::chrono::milliseconds transaction_timeout_ms,
   model::timeout_clock::duration timeout,
   model::producer_identity expected_pid) {
+    size_t tx_cache_size;
+    if (auto opt = co_await stm->tx_cache_size(); opt) {
+        tx_cache_size = opt.value();
+    } else {
+        vlog(
+          txlog.trace,
+          "can't get tx_cache_size on initializing tx.id={}",
+          tx_id);
+        co_return init_tm_tx_reply{tx_errc::not_coordinator};
+    }
+
+    if (tx_cache_size > _max_transactions_per_coordinator()) {
+        // lock is sloppy and doesn't guarantee that tx_cache_size
+        // never exceeds _max_transactions_per_coordinator
+        auto init_units = co_await stm->get_tx_thrashing_lock().get_units();
+
+        if (auto opt = co_await stm->tx_cache_size(); opt) {
```

Comment:
i am wondering why do we need sync for `tx_cache_size` and not for `oldest_tx` ?

Reply:
no particular reason, working on making it uniform